File: app/src/main/java/com/omar/retromp3recorder/app/share/SharingModule.java
package com.omar.retromp3recorder.app.share;

import io.reactivex.Completable;
import io.reactivex.Observable;

public interface SharingModule {

    Completable share();

    Observable<Event> observeEvents();

    interface Event { }

    final class SharingOk implements Event {
        public final String message;

        SharingOk(String message) {
            this.message = message;
        }
    }

    final class SharingError implements Event {
        public final String error;

        SharingError(String error) {
            this.error = error;
        }
    }
}

File: app/src/main/java/com/omar/retromp3recorder/app/share/SharingModule.kt
package com.omar.retromp3recorder.app.share

import com.github.alkurop.stringerbell.Stringer
import io.reactivex.Completable
import io.reactivex.Observable

interface SharingModule {
    fun share(): Completable
    fun observeEvents(): Observable<Event>

    sealed class Event {
        data class SharingOk internal constructor(val message: Stringer) : Event()
        data class SharingError internal constructor(val error: Stringer) : Event()
    }
}
